lib/internal/event_target.js

```diff
@@ -0,0 +1,392 @@
+'use strict';
+
+const {
+  ArrayFrom,
+  Object,
+  Map,
+  SafeWeakMap,
+  Set,
+  SymbolFor,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_THIS,
+    ERR_EVENT_RECURSION,
+  }
+} = require('internal/errors');
+const kRejection = SymbolFor('nodejs.rejection');
+
+const perf_hooks = require('perf_hooks');
+const { customInspectSymbol } = require('internal/util');
+const { inspect } = require('util');
+
+const events = new SafeWeakMap();
+const targets = new SafeWeakMap();
```

Comment:
While being weak, those two globals scares me quit a bit.
What's the memory profile of widespread usage, e.g. 1mln events/s?
Also, what's the impact on CPU?

I have a feeling that this type of implementation is performance-savvy at all.

Reply:
Right – we can use private properties now, maybe we should?

(I’m personally still a fan of plain symbols but I feel like I’m not in the majority there.)

Reply:
dw i also prefer symbols

Reply:
Yep, I have not benchmarked this. I did it this way initially as it matches up with some of the userland implementations out there. I want to add some benchmarks for this and benchmark the different approaches.

Reply:
Specifically, I'm less concerned right now with the performance as I am with the basic model and behavior.

Reply:
Ignoring the performance aspect, I think either private fields or Symbols would improve code readability :)

Reply:
> Specifically, I'm less concerned right now with the performance as I am with the basic model and behavior.

I'm concerned with adding a new global primitive that relies on global WeakMaps.

Reply:
Btw, just benchmarking the two variants (with the WeakMap vs. private properties) and the performance for both are about the same and both *completely* abysmal compared to `EventEmitter` ;-) ... I'll be working on improving that next.